src/backend/distributed/metadata/pg_get_object_address.c

```diff
@@ -0,0 +1,503 @@
+	 * type.
+	 */
+	switch (type)
+	{
+		case OBJECT_TABLE:
+		case OBJECT_SEQUENCE:
+		case OBJECT_VIEW:
+		case OBJECT_MATVIEW:
+		case OBJECT_INDEX:
+		case OBJECT_FOREIGN_TABLE:
+		case OBJECT_COLUMN:
+		case OBJECT_ATTRIBUTE:
+		case OBJECT_COLLATION:
+		case OBJECT_CONVERSION:
+		case OBJECT_STATISTIC_EXT:
+		case OBJECT_TSPARSER:
+		case OBJECT_TSDICTIONARY:
+		case OBJECT_TSTEMPLATE:
+		case OBJECT_TSCONFIGURATION:
+		case OBJECT_DEFAULT:
+		case OBJECT_POLICY:
+		case OBJECT_RULE:
+		case OBJECT_TRIGGER:
+		case OBJECT_TABCONSTRAINT:
+		case OBJECT_OPCLASS:
+		case OBJECT_OPFAMILY:
+		{
+			objnode = (Node *) name;
+			break;
+		}
+
+		case OBJECT_ACCESS_METHOD:
+		case OBJECT_DATABASE:
+		case OBJECT_EVENT_TRIGGER:
+		case OBJECT_EXTENSION:
+		case OBJECT_FDW:
+		case OBJECT_FOREIGN_SERVER:
+		case OBJECT_LANGUAGE:
+		case OBJECT_PUBLICATION:
+		case OBJECT_ROLE:
+		case OBJECT_SCHEMA:
+		case OBJECT_SUBSCRIPTION:
+		case OBJECT_TABLESPACE:
+		{
+			if (list_length(name) != 1)
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("name list length must be exactly %d", 1)));
+			}
+			objnode = linitial(name);
+			break;
+		}
+
+		case OBJECT_TYPE:
+		case OBJECT_DOMAIN:
+		{
+			objnode = (Node *) typename;
+			break;
+		}
+
+		case OBJECT_CAST:
+		case OBJECT_DOMCONSTRAINT:
+		case OBJECT_TRANSFORM:
+		{
+			objnode = (Node *) list_make2(typename, linitial(args));
+			break;
+		}
+
+		case OBJECT_PUBLICATION_REL:
+		{
+			objnode = (Node *) list_make2(name, linitial(args));
+			break;
+		}
+
+		case OBJECT_USER_MAPPING:
+		{
+			objnode = (Node *) list_make2(linitial(name), linitial(args));
+			break;
+		}
+
+		case OBJECT_DEFACL:
+		{
+			objnode = (Node *) lcons(linitial(args), name);
+			break;
+		}
+
+		case OBJECT_AMOP:
+		case OBJECT_AMPROC:
+		{
+			objnode = (Node *) list_make2(name, args);
+			break;
+		}
+
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_ROUTINE:
+		case OBJECT_AGGREGATE:
+		case OBJECT_OPERATOR:
+		{
+			ObjectWithArgs *owa = makeNode(ObjectWithArgs);
+
+			owa->objname = name;
+			owa->objargs = args;
+			objnode = (Node *) owa;
+			break;
+		}
+
+		case OBJECT_LARGEOBJECT:
+		{
+			/* already handled above */
+			break;
+		}
+
+			/* no default, to let compiler warn about missing case */
+	}
+
+	if (objnode == NULL)
+	{
+		elog(ERROR, "unrecognized object type: %d", type);
+	}
+
+	ObjectAddress addr = get_object_address(type, objnode,
+											&relation, AccessShareLock, false);
+
+	/* CITUS CODE BEGIN */
+
+	if (accessCheck)
+	{
+		Oid userId = GetUserId();
+		AclMode aclMaskResult = 0;
+		Oid idToCheck = InvalidOid;
+
+		switch (type)
+		{
+			case OBJECT_ACCESS_METHOD:
+			{
+				if (IsObjectAddressOwnedByExtension(&addr, NULL))
+				{
+					elog(ERROR,
+						 "Current user does not have required access privileges on access method %d with type %d",
+						 addr.objectId, type);
+				}
+				aclMaskResult = 1;
```

Comment:
what does `aclMaskResult = 1` mean here? Can't we use the acl macros? Definitely needs comment for all other places as well